client/allocrunner/taskrunner/identity_hook.go

```diff
@@ -30,16 +33,45 @@ type IdentitySigner interface {
 	SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error)
 }
 
+// tokenSetter provides methods for exposing workload identities to other
+// internal Nomad components.
+type tokenSetter interface {
+	setNomadToken(token string)
+}
+
 type identityHook struct {
-	tr       *TaskRunner
-	tokenDir string
-	logger   log.Logger
+	alloc      *structs.Allocation
+	task       *structs.Task
+	tokenDir   string
+	envBuilder *taskenv.Builder
+	tr         tokenSetter
+	widmgr     IdentitySigner
+	logger     log.Logger
+
+	// minWait is the minimum amount of time to wait before renewing. Settable to
+	// ease testing.
+	minWait time.Duration
+
+	stopCtx context.Context
+	stop    context.CancelFunc
 }
 
 func newIdentityHook(tr *TaskRunner, logger log.Logger) *identityHook {
+	// Create a context for the renew loop. This context will be canceled when
+	// the task is stopped or agent is shutting down, unlike Prestart's ctx which
+	// is not intended for use after Prestart is returns.
+	stopCtx, stop := context.WithCancel(context.Background())
```

Comment:
The task runner has its own shutdown context as well (for the task's own lifecycle), should we be using that here instead of the background context?

Reply:
Good catch. Looking at [`vault_hook.go`]() we use a `ShutdownHook` for that purpose, so I'll follow that pattern here.

https://github.com/hashicorp/nomad/blob/v1.6.1/client/allocrunner/taskrunner/vault_hook.go#L180-L182